pkg/component/etcd/statefulset/statefulset.go

```diff
@@ -302,11 +308,25 @@ func (c *component) waitUntilPodsHaveDesiredLabels(ctx context.Context, etcd *dr
 			if !utils.ContainsAllDesiredLabels(pod.Labels, c.values.PodLabels) {
 				return false, nil
 			}
+			// If labels were patched in the current pre-deploy flow, then wait for all pods to be ready.
+			// This is to ensure that the double rolling of the sts pods does not lead to transient quorum loss.
+			if *labelsPatched && !isPodReady(pod) {
```

Comment:
@shreyas-s-rao 
I'm concerned about the scenario where a transient quorum loss occurs. If the error reconciliation process triggers a retry later, it seems we may not recheck pod readiness if the initial readiness check fails.

Reply:
In this branch of druid, the reconciliation loop does not exit early and retry operation later. Instead, it waits for the operations to succeed or fail and then continue. So this PR fixes this issue of transient quorum loss for the happy path, which is still a large improvement over what we have right now on this branch.

In the happy path of pre-deploy, reconciler patches labels, waits for rolling, then deletes sts within the same reconciliation loop. If the reconciler errors or druid restarts during this period for any reason, then we face a transient quorum loss. But this is a rare case, and with the reconciliation design of this branch (v0.22), there's not much we can do here. Such issues don't occur on master branch since reconciliation is streamlined and made resilient to restarts by #777 .

So I would say we can merge this PR without further changes. After all, the worst case is still a transient quorum loss, which causes 1-2 minutes of downtime. And in the case of gardener, this happens within the shoot maintenance window, so the impact is minimized.

Reply:
👍